apps/yggdrasil/test/yggdrasil/user_role.exs

```diff
@@ -0,0 +1,104 @@
+defmodule Yggdrasil.UserRoleTest do
+  use ExUnit.Case, async: false
+
+  alias Yggdrasil.{Repo, User, Role, UserRole, RoleResource, Resource, Permission}
+
+  @user %{username: "tester",
+          password: "password",
+          password_confirmation: "password"}
+
+  @test_perms [:foo_perm, :bar_perm, :baz_perm]
+  @test_resources [:foo_res, :bar_res, :baz_res]
+  @test_roles [
+    %{
+      name: :foo_role,
+      resources: [
+        %{
+          name: :foo_res,
+          perms: [:foo_perm, :bar_perm]
+        },
+        %{
+          name: :bar_res,
+          perms: [:bar_perm, :baz_perm]
+        }
+      ]
+    }
+  ]
+
+  setup tags do
+    unless tags[:async] do
+      Ecto.Adapters.SQL.restart_test_transaction(Yggdrasil.Repo, [])
+    end
+
+    :ok
+  end
+
+  setup _ctx do
+    perms = Enum.map @test_perms, fn p ->
+      Repo.insert! %Permission{name: Atom.to_string(p)}
+    end
+
+    res = Enum.map @test_resources, fn r ->
+      Repo.insert! %Resource{name: Atom.to_string(r)}
+    end
+
+    roles = Enum.map @test_roles, fn tr ->
+      role = Repo.insert! %Role{name: Atom.to_string(tr.name)}
+
+      Enum.each tr.resources, fn trs ->
+        resource = Enum.find res, fn rs -> rs.name == Atom.to_string(trs.name) end
+
+        Enum.each trs.perms, fn trpm ->
+          permission = Enum.find perms, fn pm -> 
+            pm.name == Atom.to_string(trpm)
+          end
+
+          Repo.insert! %RoleResource{
+            role_id: role.id,
+            resource_id: resource.id,
```

Comment:
Good point, it would also enforce the uniqueness of the resource and perms as it should be unique.

Hmm, this cleans a few things I think, also no need to load the resource or permission when looking at `RolePermission` wouldn't it? Since the string would be in both tables? 

`RolePermission.permission` would be the permission and no need to look it up, the DB enforces the constraint so it's valid but yeah cool I like that idea.